File: packages/gavio-java/gavio-core/src/test/java/io/gavio/inspector/InspectorAgenticTest.java
package io.gavio.inspector;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.gavio.Gateway;
import io.gavio.GavioRequest;
import io.gavio.json.Json;
import io.gavio.providers.MockProvider;
import io.gavio.types.Message;
import io.gavio.types.Provider;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

/**
 * Inspector v0.7.0 endpoints (F-OBS-10 / F-DX-08 / F-DX-11 / F-DX-12): DAG,
 * sessions, stats, replay, simulate-cost, export and chain verification.
 * Mirrors the Python reference tests in {@code test_inspector_agentic.py}.
 */
class InspectorAgenticTest {

    private final HttpClient client = HttpClient.newHttpClient();
    private Gateway gateway;

    @AfterEach
    void stopServer() {
        if (gateway != null && gateway.inspector() != null) {
            gateway.inspector().stop();
        }
    }

    private String serve(CaptureMode mode) {
        gateway = Gateway.builder()
                .adapter(new MockProvider())
                .model("mock")
                .inspect(InspectorConfig.builder()
                        .enabled(true)
                        .port(0)
                        .mode(mode)
                        .unsafeContentCaptureAck(true)
                        .build())
                .build();
        return "http://127.0.0.1:" + gateway.inspector().port();
    }

    private HttpResponse<String> get(String url) throws IOException, InterruptedException {
        return client.send(HttpRequest.newBuilder(URI.create(url)).build(),
                HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> post(String url, String body) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();
        return client.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private String complete(String content, String agentId, String parentTraceId, String sessionId) {
        GavioRequest request = GavioRequest.builder()
                .messages(List.of(Message.of("user", content)))
                .model("mock")
                .provider(Provider.MOCK)
                .agentId(agentId)
                .parentTraceId(parentTraceId)
                .sessionId(sessionId)
                .build();
        return gateway.complete(request).join().traceId();
    }

    /** One orchestrator trace with two children in session s1. */
    private String[] runFamily() {
        String root = complete("orchestrate", "orchestrator", null, "s1");
        String childA = complete("sub-task a", "worker-a", root, "s1");
        String childB = complete("sub-task b", "worker-b", root, "s1");
        return new String[] {root, childA, childB};
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> listOfMaps(Object value) {
        return (List<Map<String, Object>>) value;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> map(Object value) {
        return (Map<String, Object>) value;
    }

    private static long asLong(Object value) {
        return ((Number) value).longValue();
    }

    // ---- F-OBS-10: DAG + sessions --------------------------------------------

    @Test
    void dagEndpointBuildsCallGraphWithRollups() throws Exception {
        String base = serve(CaptureMode.FULL);
        String[] family = runFamily();
        String rootId = family[0];

        Map<String, Object> dag = Json.parseObject(get(base + "/api/dag?root=" + rootId).body());
        List<Map<String, Object>> nodes = listOfMaps(dag.get("nodes"));
        Set<String> nodeIds = new TreeSet<>();
        for (Map<String, Object> node : nodes) {
            nodeIds.add((String) node.get("traceId"));
        }
        assertEquals(new TreeSet<>(Set.of(family[0], family[1], family[2])), nodeIds);

        List<Map<String, Object>> edges = listOfMaps(dag.get("edges"));
        assertEquals(2, edges.size());
        Set<String> edgePairs = new TreeSet<>();
        for (Map<String, Object> edge : edges) {
            edgePairs.add(edge.get("from") + ">" + edge.get("to"));
        }
        assertEquals(new TreeSet<>(Set.of(rootId + ">" + family[1], rootId + ">" + family[2])),
                edgePairs);

        Map<String, Object> rootNode = nodes.stream()
                .filter(n -> rootId.equals(n.get("traceId"))).findFirst().orElseThrow();
        assertEquals("orchestrator", rootNode.get("agentId"));
        Map<String, Object> subtree = map(rootNode.get("subtree"));
        assertEquals(3L, asLong(subtree.get("traces")));
        assertEquals(0L, asLong(subtree.get("errors")));
        Map<String, Object> leaf = nodes.stream()
                .filter(n -> family[1].equals(n.get("traceId"))).findFirst().orElseThrow();
        assertEquals(1L, asLong(map(leaf.get("subtree")).get("traces")));

        Map<String, Object> bySession = Json.parseObject(get(base + "/api/dag?session_id=s1").body());
        assertEquals(3, listOfMaps(bySession.get("nodes")).size());

        assertEquals(400, get(base + "/api/dag").statusCode());
        assertEquals(404, get(base + "/api/dag?root=no-such-trace").statusCode());
    }

    @Test
    void sessionsEndpointAggregatesBySession() throws Exception {
        String base = serve(CaptureMode.FULL);
        runFamily();
        complete("no session", null, null, null);

        Map<String, Object> body = Json.parseObject(get(base + "/api/sessions").body());
        List<Map<String, Object>> sessions = listOfMaps(body.get("sessions"));
        assertEquals(1, sessions.size());
        Map<String, Object> s1 = sessions.get(0);
        assertEquals("s1", s1.get("sessionId"));
        assertEquals(3L, asLong(s1.get("traces")));
        assertEquals(0L, asLong(s1.get("errors")));
        Set<String> agents = new TreeSet<>();
        for (Object agent : (List<?>) s1.get("agents")) {
            agents.add((String) agent);
        }
        assertEquals(new TreeSet<>(Set.of("orchestrator", "worker-a", "worker-b")), agents);
        String first = (String) s1.get("firstWallTimeUtc");
        String last = (String) s1.get("lastWallTimeUtc");
        assertTrue(first.compareTo(last) <= 0, "firstWallTimeUtc must not exceed lastWallTimeUtc");
    }

    // ---- F-DX-08: stats --------------------------------------------------------

    @Test
    void statsEndpointRedAggregatesAndGrouping() throws Exception {
        String base = serve(CaptureMode.FULL);
        runFamily();

        Map<String, Object> stats = Json.parseObject(get(base + "/api/stats").body());
        Map<String, Object> total = map(stats.get("total"));
        assertEquals(3L, asLong(total.get("requests")));
        assertEquals(0L, asLong(total.get("errors")));
        assertEquals(0.0, ((Number) total.get("errorRate")).doubleValue());
        assertNotNull(map(total.get("latencyMs")).get("p50"));
        assertTrue(asLong(map(total.get("tokens")).get("total")) > 0, "usage must be captured");
        assertEquals(0.0, ((Number) total.get("cacheHitRate")).doubleValue());

        Map<String, Object> grouped =
                Json.parseObject(get(base + "/api/stats?group_by=agent_id").body());
        Map<String, Object> groups = map(grouped.get("groups"));
        assertEquals(new TreeSet<>(Set.of("orchestrator", "worker-a", "worker-b")),
                new TreeSet<>(groups.keySet()));
        assertEquals(1L, asLong(map(groups.get("worker-a")).get("requests")));

        assertEquals(400, get(base + "/api/stats?group_by=nope").statusCode());
    }

    @Test
    void statsCountsPiiAndErrorsFromSummaries() {
        Map<String, Object> a = new LinkedHashMap<>();
        a.put("traceId", "a");
        a.put("status", "ok");
        a.put("latencyMs", 10);
        a.put("piiEntityTypes", List.of("EMAIL"));
        Map<String, Object> b = new LinkedHashMap<>();
        b.put("traceId", "b");
        b.put("status", "error");
        b.put("latencyMs", 30);
        b.put("piiEntityTypes", List.of("EMAIL", "IBAN"));

        Map<String, Object> total = map(
                InspectorAnalytics.buildStats(List.of(a, b), null, null).get("total"));
        assertEquals(1L, asLong(total.get("errors")));
        assertEquals(0.5, ((Number) total.get("errorRate")).doubleValue());
        assertEquals(Map.of("EMAIL", 2L, "IBAN", 1L), total.get("piiDetections"));
        assertEquals(10L, asLong(map(total.get("latencyMs")).get("p50")));
        assertEquals(30L, asLong(map(total.get("latencyMs")).get("p99")));
    }

    // ---- F-DX-11: replay --------------------------------------------------------

    @Test
    void replayRefiresTraceAndReturnsNewTraceId() throws Exception {
        String base = serve(CaptureMode.FULL);
        String traceId = complete("replay me", null, null, null);

        HttpResponse<String> replayResponse =
                post(base + "/api/replay", "{\"traceId\": \"" + traceId + "\"}");
        assertEquals(200, replayResponse.statusCode());
        Map<String, Object> replayed = Json.parseObject(replayResponse.body());
        assertEquals(traceId, replayed.get("replayOf"));
        assertNotEquals(traceId, replayed.get("traceId"));

        // The replayed call went through the live pipeline into the buffer.
        Map<String, Object> newTrace =
                Json.parseObject(get(base + "/api/traces/" + replayed.get("traceId")).body());
        assertEquals("ok", map(newTrace.get("summary")).get("status"));

        HttpResponse<String> editedResponse = post(base + "/api/replay",
                "{\"traceId\": \"" + traceId + "\", \"overrides\": "
                        + "{\"messages\": [{\"role\": \"user\", \"content\": \"edited\"}]}}");
        assertEquals(200, editedResponse.statusCode());
        Map<String, Object> edited = Json.parseObject(editedResponse.body());
        Map<String, Object> editedTrace =
                Json.parseObject(get(base + "/api/traces/" + edited.get("traceId")).body());
        Map<String, Object> start = listOfMaps(editedTrace.get("events")).stream()
                .filter(e -> "trace.start".equals(e.get("type"))).findFirst().orElseThrow();
        List<Map<String, Object>> messages = listOfMaps(map(start.get("data")).get("messages"));
        assertEquals("edited", messages.get(0).get("content"));

        assertEquals(404,
                post(base + "/api/replay", "{\"traceId\": \"no-such-trace\"}").statusCode());
        assertEquals(400, post(base + "/api/replay", "{}").statusCode());
    }

    @Test
    void replayIs403OutsideFullMode() throws Exception {
        String base = serve(CaptureMode.REDACTED);
        String traceId = complete("hi", null, null, null);
        assertEquals(403,
                post(base + "/api/replay", "{\"traceId\": \"" + traceId + "\"}").statusCode());
    }

    // ---- cost simulator -----------------------------------------------------------

    @Test
    void simulateCostRecostsTraceUnderOtherModel() throws Exception {
        String base = serve(CaptureMode.FULL);
        String traceId = complete("price this call", null, null, null);

        HttpResponse<String> response =
                get(base + "/api/simulate-cost?trace_id=" + traceId + "&model=gpt-4o");
        assertEquals(200, response.statusCode());
        Map<String, Object> simulated = Json.parseObject(response.body());
        assertEquals(traceId, simulated.get("traceId"));
        assertEquals("gpt-4o", simulated.get("simulatedModel"));
        double simulatedCost = ((Number) simulated.get("simulatedCostUsd")).doubleValue();
        assertTrue(simulatedCost > 0.0, "mock is free, gpt-4o is not");
        double delta = ((Number) simulated.get("deltaUsd")).doubleValue();
        double original = ((Number) simulated.get("costUsd")).doubleValue();
        assertEquals(simulatedCost - original, delta, 1e-9);
        assertTrue(asLong(map(simulated.get("usage")).get("totalTokens")) > 0);

        assertEquals(400, get(base + "/api/simulate-cost?trace_id=missing-both").statusCode());
    }

    // ---- F-DX-12: export -------------------------------------------------------------

    @Test
    void exportTestVectorSanitizesPii() throws Exception {
        String base = serve(CaptureMode.FULL);
        String traceId = complete("mail bob.real@corp.com about the invoice", null, null, null);

        HttpResponse<String> vectorResponse =
                get(base + "/api/traces/" + traceId + "/export?format=test-vector");
        assertEquals(200, vectorResponse.statusCode());
        Map<String, Object> vector = Json.parseObject(vectorResponse.body());
        assertTrue(((String) vector.get("id")).startsWith("exported-"));
        assertEquals("full", vector.get("mode"));
        List<Map<String, Object>> messages =
                listOfMaps(map(vector.get("request")).get("messages"));
        String content = (String) messages.get(0).get("content");
        assertFalse(content.contains("bob.real@corp.com"), "real email must be sanitized");
        assertTrue(content.contains("jan@example.com"), "synthetic fixture must replace PII");
        List<Map<String, Object>> expectedEvents = listOfMaps(vector.get("expectedEvents"));
        assertEquals("trace.start", expectedEvents.get(0).get("type"));
        assertEquals("trace.end", expectedEvents.get(expectedEvents.size() - 1).get("type"));

        HttpResponse<String> testkit =
                get(base + "/api/traces/" + traceId + "/export?format=testkit-py");
        assertEquals(200, testkit.statusCode());
        assertTrue(testkit.body().contains("GavioTestKit"));
        assertFalse(testkit.body().contains("bob.real@corp.com"));

        HttpResponse<String> testkitJava =
                get(base + "/api/traces/" + traceId + "/export?format=testkit-java");
        assertEquals(200, testkitJava.statusCode());
        assertTrue(testkitJava.body().contains("GavioTestKit"));

        assertEquals(400,
                get(base + "/api/traces/" + traceId + "/export?format=nope").statusCode());
    }

    @Test
    void exportIs403InMetadataMode() throws Exception {
        String base = serve(CaptureMode.METADATA);
        String traceId = complete("hi", null, null, null);
        assertEquals(403,
                get(base + "/api/traces/" + traceId + "/export?format=test-vector").statusCode());
    }

    // ---- F-DX-08: chain verify --------------------------------------------------------

    @Test
    void chainVerifyIs400OnLiveServer() throws Exception {
        String base = serve(CaptureMode.FULL);
        assertEquals(400, get(base + "/api/chain/verify").statusCode());
    }

    // ---- analytics unit coverage --------------------------------------------------------

    @Test
    void buildDagToleratesParentCycles() {
        Map<String, Object> a = new LinkedHashMap<>();
        a.put("traceId", "a");
        a.put("parentTraceId", "b");
        a.put("status", "ok");
        a.put("latencyMs", 1);
        a.put("costUsd", 0.0);
        Map<String, Object> b = new LinkedHashMap<>();
        b.put("traceId", "b");
        b.put("parentTraceId", "a");
        b.put("status", "ok");
        b.put("latencyMs", 1);
        b.put("costUsd", 0.0);

        Map<String, Object> dag = InspectorAnalytics.buildDag(List.of(a, b), "a", null);
        assertNotNull(dag);
        Set<String> nodeIds = new TreeSet<>();
        for (Map<String, Object> node : listOfMaps(dag.get("nodes"))) {
            nodeIds.add((String) node.get("traceId"));
        }
        assertEquals(new TreeSet<>(Set.of("a", "b")), nodeIds);
    }

    @Test
    void buildSessionsSkipsSessionlessTraces() {
        Map<String, Object> summary = new LinkedHashMap<>();
        summary.put("traceId", "a");
        summary.put("sessionId", null);
        assertEquals(List.of(), InspectorAnalytics.buildSessions(List.of(summary)));
    }
}

File: packages/gavio-js/tests/unit/inspector-agentic.test.ts
/**
 * Inspector v0.7.0 tests — DAG, sessions, stats, replay, simulate-cost, export.
 *
 * Covers F-OBS-10 (agent call graph + sessions), F-DX-11 (replay & edit-resend),
 * F-DX-08 (RED stats + hash-chain verify gate) and F-DX-12 (export trace as a
 * test case). Mirrors the Python `tests/unit/test_inspector_agentic.py`.
 */

import { describe, it, expect, afterEach } from 'vitest'
import { Gateway } from '../../src/gateway.js'
import { buildDag, buildSessions, buildStats } from '../../src/inspector/analytics.js'
import type { InspectorMode } from '../../src/inspector/index.js'
import { piiGuard } from '../../src/interceptors/pii/index.js'
import { mockProvider } from '../../src/providers/mock.js'

const gateways: Gateway[] = []

afterEach(async () => {
  for (const gw of gateways.splice(0)) await gw.inspector?.close()
})

async function serve(mode: InspectorMode = 'full'): Promise<{ gw: Gateway; base: string }> {
  const gw = new Gateway({
    model: 'mock',
    inspect: { mode, port: 0, unsafeContentCaptureAck: true },
  }).withAdapter(mockProvider())
  gateways.push(gw)
  const port = await gw.inspector!.ready()
  return { gw, base: `http://127.0.0.1:${port}` }
}

async function getJson(base: string, path: string): Promise<Record<string, unknown>> {
  const response = await fetch(base + path)
  expect(response.status).toBe(200)
  return (await response.json()) as Record<string, unknown>
}

async function post(base: string, path: string, body: unknown): Promise<Response> {
  return fetch(base + path, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  })
}

/** One orchestrator trace with two children in session s1. */
async function runFamily(gw: Gateway): Promise<[string, string, string]> {
  const root = await gw.complete({
    messages: [{ role: 'user', content: 'orchestrate' }],
    agentId: 'orchestrator',
    sessionId: 's1',
  })
  const childA = await gw.complete({
    messages: [{ role: 'user', content: 'sub-task a' }],
    agentId: 'worker-a',
    parentTraceId: root.traceId,
    sessionId: 's1',
  })
  const childB = await gw.complete({
    messages: [{ role: 'user', content: 'sub-task b' }],
    agentId: 'worker-b',
    parentTraceId: root.traceId,
    sessionId: 's1',
  })
  return [root.traceId, childA.traceId, childB.traceId]
}

interface DagResponse {
  nodes: Array<{
    traceId: string
    agentId: string | null
    subtree: { traces: number; errors: number }
  }>
  edges: Array<{ from: string; to: string }>
}

// ---- F-OBS-10: DAG + sessions -----------------------------------------------

describe('GET /api/dag', () => {
  it('builds the call graph with subtree rollups', async () => {
    const { gw, base } = await serve()
    const [rootId, childA, childB] = await runFamily(gw)

    const dag = (await getJson(base, `/api/dag?root=${rootId}`)) as unknown as DagResponse
    expect(new Set(dag.nodes.map((n) => n.traceId))).toEqual(new Set([rootId, childA, childB]))
    expect(new Set(dag.edges.map((e) => `${e.from}>${e.to}`))).toEqual(
      new Set([`${rootId}>${childA}`, `${rootId}>${childB}`]),
    )
    const rootNode = dag.nodes.find((n) => n.traceId === rootId)!
    expect(rootNode.agentId).toBe('orchestrator')
    expect(rootNode.subtree.traces).toBe(3)
    expect(rootNode.subtree.errors).toBe(0)
    const leaf = dag.nodes.find((n) => n.traceId === childA)!
    expect(leaf.subtree.traces).toBe(1)

    const bySession = (await getJson(base, '/api/dag?session_id=s1')) as unknown as DagResponse
    expect(bySession.nodes).toHaveLength(3)

    expect((await fetch(`${base}/api/dag`)).status).toBe(400)
    expect((await fetch(`${base}/api/dag?root=no-such-trace`)).status).toBe(404)
  })
})

describe('GET /api/sessions', () => {
  it('aggregates traces by session', async () => {
    const { gw, base } = await serve()
    await runFamily(gw)
    await gw.complete({ messages: [{ role: 'user', content: 'no session' }] })

    const sessions = (await getJson(base, '/api/sessions'))['sessions'] as Array<
      Record<string, unknown>
    >
    expect(sessions).toHaveLength(1)
    const s1 = sessions[0]!
    expect(s1['sessionId']).toBe('s1')
    expect(s1['traces']).toBe(3)
    expect(s1['errors']).toBe(0)
    expect([...(s1['agents'] as string[])].sort()).toEqual([
      'orchestrator',
      'worker-a',
      'worker-b',
    ])
    expect(s1['firstWallTimeUtc'] as string <= (s1['lastWallTimeUtc'] as string)).toBe(true)
  })
})

// ---- F-DX-08: stats ------------------------------------------------------------

describe('GET /api/stats', () => {
  it('serves RED aggregates and grouping', async () => {
    const { gw, base } = await serve()
    await runFamily(gw)

    const stats = await getJson(base, '/api/stats')
    const total = stats['total'] as Record<string, any>
    expect(total['requests']).toBe(3)
    expect(total['errors']).toBe(0)
    expect(total['errorRate']).toBe(0)
    expect(total['latencyMs']['p50']).not.toBeNull()
    expect(total['tokens']['total']).toBeGreaterThan(0)
    expect(total['cacheHitRate']).toBe(0)

    const grouped = await getJson(base, '/api/stats?group_by=agent_id')
    const groups = grouped['groups'] as Record<string, Record<string, unknown>>
    expect(new Set(Object.keys(groups))).toEqual(
      new Set(['orchestrator', 'worker-a', 'worker-b']),
    )
    expect(groups['worker-a']!['requests']).toBe(1)

    expect((await fetch(`${base}/api/stats?group_by=nope`)).status).toBe(400)
  })

  it('counts PII and errors from summaries (unit)', () => {
    const summaries = [
      { traceId: 'a', status: 'ok', latencyMs: 10, piiEntityTypes: ['EMAIL'] },
      { traceId: 'b', status: 'error', latencyMs: 30, piiEntityTypes: ['EMAIL', 'IBAN'] },
    ]
    const total = buildStats(summaries).total
    expect(total.errors).toBe(1)
    expect(total.errorRate).toBe(0.5)
    expect(total.piiDetections).toEqual({ EMAIL: 2, IBAN: 1 })
    expect(total.latencyMs.p50).toBe(10)
    expect(total.latencyMs.p99).toBe(30)
  })
})

// ---- F-DX-11: replay -------------------------------------------------------------

describe('POST /api/replay', () => {
  it('re-fires the trace and returns a new trace id', async () => {
    const { gw, base } = await serve()
    const response = await gw.complete({ messages: [{ role: 'user', content: 'replay me' }] })

    const replayedRes = await post(base, '/api/replay', { traceId: response.traceId })
    expect(replayedRes.status).toBe(200)
    const replayed = (await replayedRes.json()) as Record<string, string>
    expect(replayed['replayOf']).toBe(response.traceId)
    expect(replayed['traceId']).not.toBe(response.traceId)
    // The replayed call went through the live pipeline into the buffer.
    const newTrace = await getJson(base, `/api/traces/${replayed['traceId']}`)
    expect((newTrace['summary'] as Record<string, unknown>)['status']).toBe('ok')

    const editedRes = await post(base, '/api/replay', {
      traceId: response.traceId,
      overrides: { messages: [{ role: 'user', content: 'edited' }] },
    })
    expect(editedRes.status).toBe(200)
    const edited = (await editedRes.json()) as Record<string, string>
    const editedTrace = await getJson(base, `/api/traces/${edited['traceId']}`)
    const events = editedTrace['events'] as Array<{ type: string; data: Record<string, any> }>
    const start = events.find((e) => e.type === 'trace.start')!
    expect(start.data['messages'][0]['content']).toBe('edited')

    expect((await post(base, '/api/replay', { traceId: 'no-such-trace' })).status).toBe(404)
    expect((await post(base, '/api/replay', {})).status).toBe(400)
  })

  it('is 403 outside full mode', async () => {
    const { gw, base } = await serve('redacted')
    const response = await gw.complete({ messages: [{ role: 'user', content: 'hi' }] })
    expect((await post(base, '/api/replay', { traceId: response.traceId })).status).toBe(403)
  })
})

// ---- cost simulator ---------------------------------------------------------------

describe('GET /api/simulate-cost', () => {
  it('re-costs a trace under another model', async () => {
    const { gw, base } = await serve()
    const response = await gw.complete({
      messages: [{ role: 'user', content: 'price this call' }],
    })

    const simulated = (await getJson(
      base,
      `/api/simulate-cost?trace_id=${response.traceId}&model=gpt-4o`,
    )) as Record<string, any>
    expect(simulated['traceId']).toBe(response.traceId)
    expect(simulated['simulatedModel']).toBe('gpt-4o')
    expect(simulated['simulatedCostUsd']).toBeGreaterThan(0) // mock is free, gpt-4o is not
    expect(simulated['deltaUsd']).toBeCloseTo(
      simulated['simulatedCostUsd'] - simulated['costUsd'],
      8,
    )
    expect(simulated['usage']['totalTokens']).toBeGreaterThan(0)

    expect((await fetch(`${base}/api/simulate-cost?trace_id=missing-both`)).status).toBe(400)
  })
})

// ---- F-DX-12: export -----------------------------------------------------------------

describe('GET /api/traces/{id}/export', () => {
  it('sanitizes PII in the test-vector and testkit source', async () => {
    const gw = new Gateway({
      model: 'mock',
      inspect: { mode: 'full', port: 0, unsafeContentCaptureAck: true },
    })
      .withAdapter(mockProvider())
      .use(piiGuard())
    gateways.push(gw)
    const port = await gw.inspector!.ready()
    const base = `http://127.0.0.1:${port}`
    const response = await gw.complete({
      messages: [{ role: 'user', content: 'mail bob.real@corp.com about the invoice' }],
    })

    const vectorRes = await fetch(
      `${base}/api/traces/${response.traceId}/export?format=test-vector`,
    )
    expect(vectorRes.status).toBe(200)
    expect(vectorRes.headers.get('content-type')).toContain('application/json')
    const testCase = (await vectorRes.json()) as Record<string, any>
    expect(testCase['id']).toMatch(/^exported-/)
    expect(testCase['mode']).toBe('full')
    expect(testCase['interceptors']).toContain('pii_guard')
    const content = testCase['request']['messages'][0]['content'] as string
    expect(content).not.toContain('bob.real@corp.com')
    expect(content).toContain('jan@example.com')
    const expected = testCase['expectedEvents'] as Array<{ type: string }>
    expect(expected[0]!.type).toBe('trace.start')
    expect(expected[expected.length - 1]!.type).toBe('trace.end')

    const sourceRes = await fetch(
      `${base}/api/traces/${response.traceId}/export?format=testkit-py`,
    )
    expect(sourceRes.status).toBe(200)
    const source = await sourceRes.text()
    expect(source).toContain('GavioTestKit')
    expect(source).not.toContain('bob.real@corp.com')

    expect(
      (await fetch(`${base}/api/traces/${response.traceId}/export?format=nope`)).status,
    ).toBe(400)
  })

  it('is 403 in metadata mode', async () => {
    const { gw, base } = await serve('metadata')
    const response = await gw.complete({ messages: [{ role: 'user', content: 'hi' }] })
    expect(
      (await fetch(`${base}/api/traces/${response.traceId}/export?format=test-vector`)).status,
    ).toBe(403)
  })
})

// ---- F-DX-08: hash-chain verify gate + trace search -----------------------------------

describe('GET /api/chain/verify', () => {
  it('is 400 on the live server', async () => {
    const { base } = await serve()
    const response = await fetch(`${base}/api/chain/verify`)
    expect(response.status).toBe(400)
  })
})

describe('GET /api/traces?q=', () => {
  it('filters summaries by trace id prefix', async () => {
    const { gw, base } = await serve()
    const [rootId] = await runFamily(gw)

    const hits = (await getJson(base, `/api/traces?q=${rootId}`))['traces'] as Array<
      Record<string, unknown>
    >
    expect(hits.map((t) => t['traceId'])).toEqual([rootId])
    const misses = (await getJson(base, '/api/traces?q=zzzz'))['traces'] as unknown[]
    expect(misses).toHaveLength(0)
  })
})

// ---- analytics unit coverage ------------------------------------------------------

describe('analytics', () => {
  it('buildDag tolerates parent cycles', () => {
    const summaries = [
      { traceId: 'a', parentTraceId: 'b', status: 'ok', latencyMs: 1, costUsd: 0 },
      { traceId: 'b', parentTraceId: 'a', status: 'ok', latencyMs: 1, costUsd: 0 },
    ]
    const dag = buildDag(summaries, 'a')!
    expect(new Set(dag.nodes.map((n) => n.traceId))).toEqual(new Set(['a', 'b']))
  })

  it('buildSessions skips sessionless traces', () => {
    expect(buildSessions([{ traceId: 'a', sessionId: null }])).toEqual([])
  })
})
